### DISCUSSION TITLE
Make rc2nix stable

### DISCUSSION THREAD

**Discussion by luochen1990** [nix-community/plasma-manager#337] [Ideas]
Hi, I find that currently rc2nix is not stable, which means, when I run rc2nix and activate it and run rc2nix again, things changed.

Here is the diff between run rc2nix first time and second time:

![image](https://github.com/user-attachments/assets/461da47a-875a-46b3-b105-ed59bc332f6b)

![image](https://github.com/user-attachments/assets/4fa6b26f-9084-4a9a-b847-9c5b3c939098)

I hope to make it stable. And so have several questions:

1. I find that there is also `KConfManager.read()` in `write_config.py` which parse config files just as what `rc2nix.py` did, is it possible to lift the config file parsing logic into a common utils to reuse them?   (we can also do similar thing for `unparse`  so that we can add unit tests to make sure they are inverse functions of each other.)
2. Is this a bug?  I need some detailed explanation for the diff in the pictures above.

**Comment by HeitorAugustoLN** [comment]
I am thinking of moving every script to a single cli program, since I'm going to be adding another script for kxmlgui files, so they can also share utilities, because i think KConfManager.read() could be actually used for rc2nix, but I will need to look into it.

**Comment by robert-elles** [comment]
Hey,
as a workaround I have a script that validates the syntax of the output that rc2nix produces and removes the problematic lines:
```
file=~/yourfile/plasma_rc2nix.nix

rc2nix > $file

while true; do
    result=$(nix-instantiate --parse $file -v 0 --log-format internal-json 2>&1)
    result=$(echo "$result" | sed 's/.*@nix.*@nix//')
    result=$(echo "$result" | sed 's/^@nix//')
    
    line=$(echo "$result" | jq -r '.line' | sed '1d')

    if [ -z "$line" ]; then
        break
    fi

    echo "removing line: $line"
    sed -i "${line}d" $file
done
```
